File: src/main/java/br/com/blz/testjava/business/domain/Inventory.java
package br.com.blz.testjava.business.domain;

import java.util.List;

import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.OneToMany;
import javax.persistence.Table;
import javax.persistence.Transient;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Builder
@NoArgsConstructor
@AllArgsConstructor
@Data
@Entity
@Table(name = "inventory")
public class Inventory {
	
	@Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
	@Column(name = "id")
	private Long id;
	
	@Transient
	private int quantity;
	
	@OneToMany(cascade = CascadeType.ALL)
	private List<Warehouse> warehouses;
}

File: src/main/kotlin/com/boticario/testkotlin/domain/entity/Inventory.kt
package com.boticario.testkotlin.domain.entity

import java.io.InvalidObjectException

data class Inventory(
        val warehouses: MutableList<Warehouse> = mutableListOf(),
        val quantity: Int = warehouses.sumOf { it.quantity }
) {
    init {
        if(quantity != warehouses.sumOf { it.quantity })
            throw InvalidObjectException("Wrong inventory quantity!")
    }
}
